### ISSUE TITLE
REGR: Accessing dict in JITed code in 1.11

### ISSUE THREAD

**Issue by twoertwein** [pytorch/pytorch#74056]
### 🐛 Describe the bug

The following code worked with pytorch 1.10.2. Weirdly, there are many small independent (and sometimes seemingly unrelated) changes that make this error disappear.

```py
from typing import Final

import torch


class LinearActivation(torch.nn.Module):
    def forward(
        self, x: torch.Tensor, meta: dict[str, torch.Tensor]
    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
        meta = meta.copy()
        meta["meta_y_hat"] = x
        # return x, meta # would make it work
        return meta["meta_y_hat"], meta  # JIT claims it errors here


class Test(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.flag: Final = ""
        self.activation = LinearActivation()

    def forward(
        self, x: torch.Tensor, meta: dict[str, torch.Tensor]
    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
        meta = meta.copy()
        if self.flag != "":  # this branch should not even be compiled
            # assert False # would make it work
            score = x[:, -1:]
            x, meta = self.activation(
                x[:, :, :-1],  # replacing this with x, would make it work
                meta,
            )
            x = torch.cat((x, score), dim=1)  # removing this line makes it work
        else:
            x, meta = self.activation(x, meta)
        meta["meta_y_hat"] = x  # removing this line makes it work
        return meta["meta_y_hat"], meta


if __name__ == "__main__":
    model = torch.jit.script(Test())
    x, xs = model.forward(torch.ones(10, 10), {})
```

```
python test.py
Traceback (most recent call last):
  File "test.py", line 43, in <module>
    x, xs = model.forward(torch.ones(10, 10), {})
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "test.py", line 36, in forward
            x = torch.cat((x, score), dim=1)  # removing this line makes it work
        else:
            x, meta = self.activation(x, meta)
                      ~~~~~~~~~~~~~~~ <--- HERE
        meta["meta_y_hat"] = x  # removing this line makes it work
        return meta["meta_y_hat"], meta
  File "test.py", line 13, in forward
        meta["meta_y_hat"] = x
        # return x, meta # would make it work
        return meta["meta_y_hat"], meta  # JIT claims it errors here
               ~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: KeyError: meta_y_hat
```

### Versions

Collecting environment information...
PyTorch version: 1.11.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.7 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~16.04) 9.4.0
Clang version: 3.8.0-2ubuntu4 (tags/RELEASE_380/final)
CMake version: version 3.15.1
Libc version: glibc-2.23

Python version: 3.9.9+ (heads/3.9:86c1265, Nov 27 2021, 15:32:47)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-4.13.0-36-generic-x86_64-with-glibc2.23
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 440.64
cuDNN version: /usr0/local/cuda-9.0/lib64/libcudnn.so.7.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.931
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.11.0+cu102
[pip3] torchvision==0.12.0
[conda] Could not collect

**Comment by davidberard98** [comment]
Thanks for the report @twoertwein - I'm taking a look at why this is happening.

FYI, I think the usage of `Final` is incorrect in your example.

I tried with the following and it succeeds.
```python
...
class Test(torch.nn.Module):
    __constants__ = ["flag"]
    def __init__(self):
        super().__init__()

        self.flag = ""
        self.activation = LinearActivation()
...
```

or alternatively

```python
class Test(torch.nn.Module):
    flag: Final[str]
    def __init__(self): ...
```

afaik

**Comment by davidberard98** [comment]
bisected to https://github.com/pytorch/pytorch/pull/70202

**Comment by twoertwein** [comment]
Thank you for finding the cause and fixing it!

> FYI, I think the usage of `Final` is incorrect in your example.

Thank you, I wasn't aware of that: that's how Final is typically used with other type checkers such as mypy and pyright.

**Comment by davidberard98** [comment]
Should be fixed by #74794
